File: 백준/Gold/17298. 오큰수/오큰수.py
from collections import deque
import sys
input = sys.stdin.readline
n = int(input())
a = list(map(int, input().split()))

ans = [-1] * n
stack = deque()

for i in range(n):
  # top과 현재 숫자 비교해 해당 숫자의 인덱스에 현재 숫자(오큰수) 넣기
  while stack and (stack[-1][0] < a[i]):
    val, idx = stack.pop()
    ans[idx] = a[i]
  stack.append([a[i], i])

print(*ans)

File: 백준/Gold/17298. 오큰수/오큰수.js
const fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
function sol(n, arr) {
  let res = new Array(n).fill(-1);

  stack = [];
  for (let i = 0; i < arr.length; i++) {
    let cur = arr[i];

    while (stack.length && stack.at(-1)[1] < cur) {
      res[stack.pop()[0]] = cur;
    }
    stack.push([i, cur]);
  }
  return res.join(" ");
}

let arr = input[1].split(" ").map((item) => +item);
console.log(sol(+input[0], arr));